tests/semantic_analyser.cpp

```diff
@@ -2451,6 +2451,45 @@ kprobe:f { @ = hist(5); print((1, (uint16)@)); }
 )");
 }
 
+TEST(semantic_analyser, castable_map_missing_feature)
+{
+  test_error("BEGIN { @a = count(); print(@a); }",
```

Comment:
I don't think `print(@)` forces a collapse on percpu type into scalar. Rather, it's an async map print. For example, compare outputs of these two:

```
sudo bpftrace -e 'BEGIN { @ = count(); print(@) }' -dcodegen
```

```
sudo bpftrace -e 'BEGIN { @ = count(); print((int64)@) }' -dcodegen
```

Reply:
Ah ok, I'll fix.

Reply:
As per my updated comment, it seems that checking for this case in semantic analyser would involve a lot of ugly code so I just dropped in a check on the codegen path, which isn't great because we lose the line number. But I don't think all the extra code in semantic analyser is worth it to detect this feature on older kernels. WDYT?

Reply:
What was the complication? 5.15 is still somewhat new - I expect it'll be around for 2-4 more years.

And returning error from codegen is not very nice, so it'd be preferable to find another way.

Reply:
@danobi Here is what this looks like checked in semantic analyser. It's not too bad but my main concern is adding another property of the Map ast type AND potentially missing setting this property to false on a case I didn't account for because the default needs to be 'true'. https://github.com/bpftrace/bpftrace/pull/3752

Reply:
Hmm, agreed it's not the prettiest. But given it's allowlist style, actually not too bad? If you miss a case for some new feature, then user is blocked from doing something that should work. Which is a better outcome than being allowed to do something that won't work.

That seems like something that would be more easily surfaced in tests, especially once we add an "oldest supported kernel" vmtest job.

Long term maybe we can move all the feature checking out into its own pass (cc @amscanne)? That could help contain the complexity and make details like this more obvious.

Reply:
> That seems like something that would be more easily surfaced in tests, especially once we add an "oldest supported kernel" vmtest job.

I was thinking the same thing. I tried to run the runtime test locally with vmtest but they all kept timing out - I was probably doing something wrong.

> Long term maybe we can move all the feature checking out into its own pass

Makes sense.

So would you rather I abandon this PR for the other one? My slight preference is for this one to reduce complexity in semantic analyser and know that I'm always erroring when we're going to actually call the bpf function but I'll have some solace if we land the vmtest stuff and we can run the runtime tests on a 5.15 kernel.

Reply:
I prefer the other PR (ie doing it in semantic analyser). Codegen is not designed to return errors. And we can emit better diagnostics in semantic analyser. Imagine having a large script with lots of operations on a map. The codegen exception would not be super helpful.

Maybe just document the new map field with a comment and tag an issue that moves all the feature check stuff out into its own pass?

I could see it being a lot simpler in its own pass. For example, you don't need to modify the ast - you could have an `std::set <Node *> maps_that_are_read`  that keeps track of the information. Doing that kind of collection in semantic analyser is more annoying.